metadata/ota_metadata/metadata_gen.py

```diff
@@ -11,15 +13,53 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
-
-#!/usr/bin/env python3
-
 import os
-from hashlib import sha256
 import argparse
 import pathlib
+import zstandard
 import igittigitt
+from hashlib import sha256
+
+ZSTD_COMPRESSION_EXTENSION = "zst"
+ZSTD_COMPRESSION_LEVEL = 10
+ZSTD_MULTITHREADS = 2
+CHUNK_SIZE = 4 * (1024**2)  # 4MiB
+
+
+def zstd_compress_file(
+    cctx: zstandard.ZstdCompressor,
+    src_fpath: str,
+    dst_fpath: str,
+    *,
+    cmpr_ratio: float,
+    filesize_threshold: int,
+) -> bool:
+    if (src_size := os.path.getsize(src_fpath)) < filesize_threshold:
+        return False  # skip file with too small size
+
+    try:
+        with open(src_fpath, "rb") as src_f, open(dst_fpath, "wb") as dst_f:
+            with cctx.stream_writer(dst_f, size=src_size) as compressor:
+                while data := src_f.read(CHUNK_SIZE):
+                    compressor.write(data)
+    except Exception:  # cleanup on failure
+        try:
+            os.remove(dst_fpath)
+        except OSError:
+            pass
+        return False
```

Comment:
What kind of error do we face?
It is my opinion that if we face a unintended error, it should be raised.

Reply:
Ah you mean that we should interrupt the whole process if any of the compression failed!

Updated in [51a4ec1](https://github.com/tier4/ota-metadata/pull/12/commits/51a4ec1f55496a7c246ca27993e141d1a47edee5).

Reply:
Yes, `unintended` should be debugged and fixed. So we need to know `unintended` behavior.